File: 116. Populating Next Right Pointers in Each Node.cpp
/**
 * Definition for binary tree with next pointer.
 * struct TreeLinkNode {
 *  int val;
 *  TreeLinkNode *left, *right, *next;
 *  TreeLinkNode(int x) : val(x), left(NULL), right(NULL), next(NULL) {}
 * };
 */
class Solution {
public:
    void connect(TreeLinkNode *root) {
        if (!root) return;
        queue<TreeLinkNode*> q;
        q.push(root);
        int cur = 1, next = 0;
        while (!q.empty()) {
            TreeLinkNode* top = q.front();
            q.pop();
            cur--;
            if (cur > 0) {
                top->next = q.front();
            }

            if (top->left) q.push(top->left), next++;
            if (top->right) q.push(top->right), next++;

            if (cur == 0) {
                cur = next;
                next = 0;
            }
        }
    }
};

File: java/116. Populating Next Right Pointers in Each Node.java
/**
 * Definition for binary tree with next pointer.
 * public class TreeLinkNode {
 *     int val;
 *     TreeLinkNode left, right, next;
 *     TreeLinkNode(int x) { val = x; }
 * }
 */
// O(n)
public class Solution {
    public void connect(TreeLinkNode root) {
        TreeLinkNode cur, nextHead;
        nextHead = root;
        cur = null;
        while (nextHead != null) {
            if (cur == null) {
                cur = nextHead;
                nextHead = cur.left;
            } else {
                cur.left.next = cur.right;
                if (cur.next != null) {
                    cur.right.next = cur.next.left;
                }
                cur = cur.next;
            }
        }
    }
}

/**
 * Definition for binary tree with next pointer.
 * public class TreeLinkNode {
 *     int val;
 *     TreeLinkNode left, right, next;
 *     TreeLinkNode(int x) { val = x; }
 * }
 */
// recursive O(n)
public class Solution {
    public void connect(TreeLinkNode root) {
        if (root == null || root.left == null) {
            return;
        }
        root.left.next = root.right;
        if (root.next != null) {
            root.right.next = root.next.left;
        }
        connect(root.left);
        connect(root.right);
    }
}
